packages/vite/src/node/server/middlewares/static.ts

```diff
@@ -141,17 +146,24 @@ export function serveRawFsMiddleware(
         return
       }
 
-      url = url.slice(FS_PREFIX.length)
-      if (isWindows) url = url.replace(/^[A-Z]:/i, '')
+      let newPathname = pathname.slice(FS_PREFIX.length)
+      if (isWindows) newPathname = newPathname.replace(/^[A-Z]:/i, '')
 
-      req.url = encodeURIComponent(url)
+      url.pathname = encodeURIComponent(newPathname)
+      req.url = url.href.slice(url.origin.length)
       serveFromRoot(req, res, next)
     } else {
       next()
     }
   }
 }
 
+function getDecodedPathname(url: URL) {
+  return url.pathname.includes('%')
+    ? decodeURIComponent(url.pathname)
+    : url.pathname
+}
```

Comment:
I think we should remove this helper and directly call `decodeURIComponent(url.pathname)` in place. If checking for `'%'` is a good optimization, then that should be done by `decodeURIComponent` internally.

Reply:
It does have a slightly better perf (maybe because of the native call overhead). But since it's a micro optimization and it is not much worth with our usecase I've removed this 👍